### PR TITLE
Memory management trace

### PR DESCRIPTION
**Pull request by sagarbhathwar**
@student-of

### CODE REVIEW

File: gc.ts [hunk 1]
```diff
@@ -380,20 +381,63 @@ export class MnS<A extends MarkableAllocator> {
...
-          throw new Error("TODO: trace string");
-        }
+          // Note(sagar): Memory layout is abstracted by allocator
+          // childPtr always points to start of data, not header
+          for(let dataPtr = childPtr; dataPtr !== childPtr + listLength * 4n; dataPtr += 4n) {
+            const elementValue = this.getField(dataPtr);
+
+            if(isPointer(elementValue)) {
+              const fieldPointerValue = extractPointer(elementValue);
+              if(!this.isMarked(fieldPointerValue)) {
```
Comment (student-of) [inline gc.ts]: `None` is represented as `i32.const 0`. Need to insert a check so we don't accidentally trace a bad pointer.

File: gc.ts [hunk 2]
```diff
@@ -380,20 +381,63 @@ export class MnS<A extends MarkableAllocator> {
...
+            while(currListAddr !== 0n) { // Not none
+
+              const key = this.getField(currListAddr);
+              const value = this.getField(currListAddr + 4n);
+              if(isPointer(key)) {
+                worklist.push(key);
+              }
+              if(isPointer(value)) {
+                worklist.push(value);
+              }
```
Comment (student-of) [inline gc.ts]: Ditto: `None` check

File: gc.ts [hunk 3]
```diff
@@ -380,20 +381,63 @@ export class MnS<A extends MarkableAllocator> {
...
+          // Layout: [32-bit TAG_LIST, 32-bit <length>, 32-bit <capacity>, data...]
+
+          // Extract value at childPtr + 4. Assumed to be a primitive value
+          const listLength = childSize;
+          
+          // Sanity check, just-in-case
+          // NOTE(sagar): probably not necessary
+          if(isPointer(listLength)) {
+            throw new Error("Pointer value stored in the place of list length");
+          }
```
Comment (student-of) [inline gc.ts:393]: You'll need to check with the List team or check `compiler.ts`. I'm not sure if the length is a tagged value or not

File: gc.ts (hunk 1 above)
Comment (student-of) [reply gc.ts]: Also, can you add that check for tracing fields with `TAG_CLASS` ?

File: gc.ts (hunk 3 above)
Comment (student-of) [reply gc.ts:393]: If it's tagged, you'll need to shift it 1 bit right.

**Comment by student-of** [comment]
Also, tests would be nice.

File: gc.ts (hunk 3 above)
Comment (sagarbhathwar) [reply gc.ts:393]: @student-of It's an untagged value

**Pull request merged**